### PR TITLE
Support nd array type

### LINKED ISSUE #3150: Support `numpy.typing.NDArray` and similar in `st.from_type()`
I found some notes [about `numpy.typing`](https://numpy.org/devdocs/reference/typing.html) with @rsokl  a while ago, and thought it worth an issue.  @honno might also be interested.

- [x] [`ArrayLike`](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike)... except that this is a `Union`, and we can't register unions
- [ ] [`DTypeLike`](https://numpy.org/devdocs/reference/typing.html#numpy.typing.DTypeLike)... also a `Union`
- [ ] [`NBitBase`](https://numpy.org/devdocs/reference/typing.html#numpy.typing.NBitBase) and specific-width subclasses... meaningless on it's own; would need to register strategies for the generics that take them, i.e. `np.floating`, `np.integer`, etc.
- [x] [`NDArray`](https://numpy.org/devdocs/reference/typing.html#numpy.typing.NDArray) - finally, something we can handle!  Generic class representing an `ndarray`, takes a dtype specifier.

Maybe we can do something about the `Union` limitation by understanding `TypeAlias`?  I don't think this information actually exists at runtime, though if it did we could also get better reprs out of it.  As a related question we thought that our registration logic for `TypeVar` *might* be too permissive.

It would also be nice to distinguish `shape: Tuple[int, ...]`, so that we could generate only positive ints, length-bounds, etc. (likely just by using the `array_shapes` strategy).  However, making the type resolution logic name-aware would open a *huge* can of worms and complicate the mental model enormously.  If we support this, it should be via the explicit `Shape: TypeAlias = Tuple[int, ...]` and `st.register_type_strategy(Shape, npst.array_shapes())` instead.

### PR DESCRIPTION
**Pull request by rsokl**
This is a replacement of #3451 and closes #3150

### CODE REVIEW

File: hypothesis-python/tests/cover/test_lookup.py [hunk 1]
```diff
@@ -34,6 +34,12 @@
 from tests.common.debug import assert_all_examples, find_any, minimal
 from tests.common.utils import fails_with, temp_registered
 
+if sys.version_info[:2] < (3, 9):
+    try:
+        from numpy.typing import NDArray
+    except ImportError:
+        NDArray = None
+
```
Comment (rsokl) [inline hypothesis-python/tests/cover/test_lookup.py:42]:
```suggestion
if sys.version_info[:2] < (3, 9):
    try:
        from numpy.typing import NDArray
    except ImportError:
        NDArray = None
else:
    NDArray = None
```

**Comment by Zac-HD** [comment]
Closing this as stale 😥

**Comment by Cheukting** [comment]
@Zac-HD any chance I can give this a try again?

**Comment by Zac-HD** [comment]
Go for it!

**Pull request closed without merging**